### DISCUSSION TITLE
Wish to use Table Parameters in GAUGE Steps but no clue

### DISCUSSION THREAD

**Discussion by firos-mailbox** [getgauge/gauge#1919] [General]
Hi All,

I wish to use Table Parameters in GAUGE Steps.

i.e. "Table parameters are used when a step is executed for multiple values in a table. For a step with an inline table parameter, the entire table becomes the parameter value."

The first line (in the above quoted text) makes me think that the step will be executed as many times as the number of entries in the table passing each column's value as argument to the glue language by the GAUGE framework.

But the second line makes me think that the step will be executed once with the entire table passed as argument and thus it is the job of the step implementation to fetch each entries from the table and use it. 

My first assumption is more justified given that GAUGE  is a framework and thus that will do the iterative thing i.e. taking each entries in the table and invoking the step. I'm less clear on this front.

The difficulty is I'm seeing nowhere examples even in the official documentation.

In the official documentation, we could see this:
* Create the following projects
     |project name| username |
     |------------|----------|
     | Gauge java | Daredevil|
     | Gauge ruby | Iron Fist|

but nowhere could be seen how it is implemented in the step file.

Note: I'm using Python as my glue language.

Could someone who have more insight into this shed some light? Any help is much appreciated. Thanks.

-
Firos

**Comment by sriv** [comment]
hi Firos, both assumptions are possible  depending on how you use the table. If you run `gauge init python` you get a sample project that has a scenario like:

```
## Vowel counts in multiple word

This is the second scenario in this specification

Here's a step that takes a table

* Almost all words have vowels
     |Word  |Vowel Count|
     |------|-----------|
     |Gauge |3          |
     |Mingle|2          |
     |Snap  |1          |
     |GoCD  |1          |
     |Rhythm|0          |
```

In this case the step syntax is `Almost all words have vowels <table>` (note how the table is passed as an argument.) In this step implementation, the code is expected to consume this table as a whole and work with it. 

```
@step("Almost all words have vowels <table>")
def assert_words_vowel_count(table):
    actual = [str(number_of_vowels(word)) for word in table.get_column_values_with_name("Word")]
    expected = [str(count) for count in table.get_column_values_with_name("Vowel Count")]
    assert expected == actual
```

If however the scenario were rewritten as :


```
# An example specification

     |Word  |Vowel Count|
     |------|-----------|
     |Gauge |3          |
     |Mingle|2          |
     |Snap  |1          |
     |GoCD  |1          |
     |Rhythm|0          |


## Vowel counts in multiple word

This is the second scenario in this specification

Here's a step that takes a table

* The word <Word> has <Vowel Count> vowels.
```

Gauge will read the table, parse and extract each column value as a variable, iterate over the rows and pass these variables to all steps that refer to it in the scenario. 

Hope this helps
